app/src/server/api/external/v1Api.router.ts

```
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
-            model: inputPayload.model,
-            inputTokens,
-            outputTokens,
-            cost,
-          },
-        });
+        try {
+          completion = await getCompletion2(typedFT, inputPayload);
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get fine-tune completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+          });
+        }
+      } else {
+        try {
+          completion = await getOpenaiCompletion(key.projectId, {
+            ...inputPayload,
+            // @ts-expect-error gotta love overloads
+            stream: input.stream,
+          });
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get OpenAI completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+            cause: error as Error,
+          });
+        }
+      }
 
-        return completion;
-      } catch (error: unknown) {
-        console.error(error);
-        throw new TRPCError({
-          message: `Failed to get completion: ${(error as Error).message}`,
-          code: "BAD_REQUEST",
+      const logRequest = ctx.headers["op-log-request"] === "true";
+      let tags: Record<string, string> = {};
+      if (ctx.headers["op-tags"]) {
+        try {
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
+        } catch (error: unknown) {
+          // ignore
+        }
+      }
+
+      if (completion instanceof Stream) {
+        // split stream to avoid locking the read mutex
+        const [recordStream, outputStream] = completion.tee();
+        void captureRecordUsage({
+          projectId: key.projectId,
+          inputPayload,
+          completion: recordStream,
+          logRequest,
+          fineTune,
+          tags,
         });
+        return outputStream.toReadableStream();
       }
+
+      void recordUsage({
```

Let's add a `.catch(e => captureException(e))` or similar to all of these so see in Sentry if these are breaking. We should strive to never silently swallow errors.

---

Also let's put this inside an `else`, so it's clear that this is stuff that happens if we're *not* streaming. I missed that on my first read-through and got confused.